imap_processing/hi/l1a/science_direct_event.py

```diff
@@ -166,52 +161,80 @@ def break_into_bits_size(binary_data: str):
     return [binary_data[i : i + 48] for i in range(0, len(binary_data), 48)]
 
 
-def create_dataset(de_data_list: list):
+def create_dataset(de_data_list: list, packet_met_time: list):
     """Create xarray dataset.
 
     Parameters
     ----------
     de_data_list : list
         Parsed direct event data list
+    packet_met_time : list
+        List of packet MET time
 
     Returns
     -------
     dataset: xarray.Dataset
         xarray dataset
     """
     # These are the variables that we will store in the dataset
-    met_subseconds = []
-    met_seconds = []
+    de_met_time = []
     esa_step = []
     trigger_id = []
     tof_1 = []
     tof_2 = []
     tof_3 = []
     de_tag = []
-    for event in de_data_list:
-        # TODO: how to check for fist event?
-        # metaevent.
-        # How to handle if first event is not
-        # metaevent? This means that current
-        # data file started with direct event.
-        # metaevent is a way to store information
-        # about bigger portion of time information. Eg.
-        # metaevent stores information about, let's say
-        # "20240319T09:30:01". Then direct event time
-        # tag stores information of time ticks since
-        # that time. Then we use those two to combine and
-        # get exact time information of each event.
+    ccsds_met = []
+
+    # How to handle if first event is not metaevent? This
+    # means that current data file started with direct event.
+    # Per Paul, log a warning and discard all direct events
+    # until we see next metaevent because it could mean
+    # that the instrument was turned off during repoint.
+
+    # Find the index of the first occurrence of the metaevent
+    first_metaevent_index = next(
+        (i for i, d in enumerate(de_data_list) if d.get("start_bitmask_data") == 0),
+        None,
+    )
+
+    if first_metaevent_index is None:
+        return None
+    elif first_metaevent_index != 0:
+        # Discard all direct events until we see next metaevent
+        # TODO: log a warning
+        de_data_list = de_data_list[first_metaevent_index:]
+        packet_met_time = packet_met_time[first_metaevent_index:]
+
+    for index, event in enumerate(de_data_list):
         if event["start_bitmask_data"] == 0:
```

Comment:
Is a start_bit_mask_data = 0 an indication that the current item in the list is a metaevent? I might be misinterpreting this, but it looks like the metaevents are separate dictionaries in the list? If that's the case and there should be a metaevent associated with each event, would it be easier to merge the metaevent and event into one dictionary?

Reply:
Yes. start_bit_mask_data = 0 indicates that it's a metaevent. metaevent and direct event are both 48-bits and but the first two bits indicates which type it is. Based on type, it needs to parse 48-bits differently. I had to keep them in the same list because order is important here. Every direct event will start with a metaevent and those direct events uses time information from the metaevent to build their full time information.